src/cpp/src/cache_manager.hpp

```diff
@@ -15,40 +15,140 @@ class CacheManager {
                 std::memset(value_cache.data(), 0, value_cache.get_byte_size());
 
                 m_key_cache.emplace_back(key_cache);
                 m_value_cache.emplace_back(value_cache);
+
+                update_request_tensor(decoder_layer_id);
             }
         } else {
             auto remote_context = m_core.get_default_context(device_name);
             for (size_t decoder_layer_id = 0; decoder_layer_id < m_device_config.get_num_layers(); ++decoder_layer_id) {
-                ov::Tensor key_cache = remote_context.create_tensor(device_config.get_cache_precision(),
-                                                                    device_config.get_key_cache_shape());
-                ov::Tensor value_cache = remote_context.create_tensor(device_config.get_cache_precision(),
-                                                                      device_config.get_value_cache_shape());
+                ov::Tensor key_cache = remote_context.create_tensor(m_device_config.get_cache_precision(),
+                                                                    key_cache_shape);
+                ov::Tensor value_cache = remote_context.create_tensor(m_device_config.get_cache_precision(),
+                                                                      value_cache_shape);
 
                 m_key_cache.emplace_back(key_cache);
                 m_value_cache.emplace_back(value_cache);
+
+                update_request_tensor(decoder_layer_id);
+            }
+        }
+    }
+
+    void increase_cache(size_t num_kv_blocks) {
+        OPENVINO_ASSERT(num_kv_blocks > m_num_allocated_kv_blocks);
+        ov::Shape new_value_cache_shape = set_first_dim_and_make_static(m_device_config.get_value_cache_shape(), num_kv_blocks);
+        ov::Shape new_key_cache_shape = set_first_dim_and_make_static(m_device_config.get_key_cache_shape(), num_kv_blocks);
+
+        const std::string device_name = m_device_config.get_device();
+        ov::Coordinate start_key{0,0,0,0};
+        ov::Coordinate start_value{0,0,0,0};
+
+        if (device_name.find("GPU") == std::string::npos) {
+            for (size_t decoder_layer_id = 0; decoder_layer_id < m_device_config.get_num_layers(); ++decoder_layer_id) {
+                ov::Coordinate end_key(m_key_cache[decoder_layer_id].get_shape());
+                ov::Coordinate end_value(m_value_cache[decoder_layer_id].get_shape());
+
+                ov::Tensor key_cache(m_device_config.get_cache_precision(), new_key_cache_shape);
+                ov::Tensor value_cache(m_device_config.get_cache_precision(), new_value_cache_shape);
+                
+                // copy current cache data
+                ov::Tensor dst_key_roi(key_cache, start_key, end_key);
+                ov::Tensor dst_value_roi(value_cache, start_value, end_value);
+                m_key_cache[decoder_layer_id].copy_to(dst_key_roi);
+                m_value_cache[decoder_layer_id].copy_to(dst_value_roi);
+
```

Comment:
`allocate_cache_if_needed` and `increase_cache` are quite similar and the only difference are these 4 lines where we need to copy initial cache (if it was allocated before)

maybe we need to merge these 2 functions? 
- allocate tensors with current size
- end roi by default points to beginning of tensor. It means we don't need to perform copy as start_roi == end_roi
- if cache is allocated before, set end_roi, copy from start_roi to end_roi
- from end_roi to end_of_tensor perform memset with zeros

Reply:
> end roi by default points to beginning of tensor. It means we don't need to perform copy as start_roi == end_roi

I used the condition for copying `m_key_cache.size() > decoder_layer_id` as before first allocation `m_key_cache` and `m_value_cache` are empty vectors. The rest implemented as you suggested, please take a look.